Adds cache option to uplinks

silkentrance opened the pull request:
Open to discussion. Please review.

*WIP*: needs tests

[GH-131] add cache option to uplinks

jachstet-sea commented on lib/storage.js:
```diff
@@ -246,7 +246,11 @@ Storage.prototype.get_tarball = function(name, filename) {
 
-    var savestream = self.local.add_tarball(name, filename)
+    var savestream
+    var docache = !!uplink.cache
```
Just curious: Any reason to have a double negation here?

silkentrance replied:
Yes, you are correct, this is rather unneccessary. Out of habit, I think...

juanpicado commented:
I'm considering to merge this one, but I need a rebase from master and unit testing. Could you take a look #142 @silkentrance  ?

silkentrance commented:
@juanpicado Will do.

juanpicado commented:
@silkentrance sure, no problem, take your time 👍

silkentrance commented:
I also fixed the existing integration tests, namely I adjusted the sinopia-test-1.2.3.tgz package to match the requirements, renaming everything to verdaccio.

I also added test/integration/config_nocache.yaml and test/integration/test_nocache.pl.

I think that these integration tests need to be re-implemented :grin:

The configuration uplink caching behavior is tested in test/unit/config.js.

juanpicado commented:
Awesome, I'll review soon 👍

silkentrance commented:
ATM I am working on implementing the functional tests, but this may take a while. I need to get my head around the test cases first.

silkentrance commented:
@juanpicado I am having a hard time with the proper configuration of server3 (config-3.yaml) in test/functional. It will always fail with a 404 on get_package.

While the integration test works as expected, I cannot get it to work in the functional tests.

Maybe you can give me some hints on how to use the functional server test API?

juanpicado commented on lib/storage.js:
```diff
@@ -278,10 +278,14 @@ class Storage {
       }
-      let savestream = self.local.add_tarball(name, filename);
+      let savestream = null;
+      if (uplink.config.cache == true) {
```
Do we need `==` for true?

juanpicado commented on test/functional/gh131.js:
```diff
@@ -0,0 +1,90 @@
+const PKG_NOCACHE = 'pkg_nocache';
+const PKG_NOCACHE2 = 'pkg_nocache2';
+
+function iscached(pkgname, tarballname) {
```
camelCase here please, `isCache`.  The same in other functions. I'll enable camelCase eslint rule soon.

juanpicado commented on test/functional/gh131.js:
```diff
@@ -0,0 +1,90 @@
+  });
+
+  before(function() {
+    let pkg = require('./lib/package')(PKG_NOCACHE2);
```
const if the variable does not change

juanpicado commented on test/functional/lib/startup.js:
```diff
@@ -41,4 +42,5 @@ module.exports.start = function(dir, conf) {
 process.on('exit', function() {
   if (forks[0]) forks[0].kill();
```
Let's use `curly` braces, I'll enable that rule soon as well.

juanpicado commented on test/unit/config.js:
```diff
@@ -0,0 +1,17 @@
+'use strict';
+
+let assert = require('assert');
```
`const` for imports

juanpicado commented on test/unit/partials/config.js:
```diff
@@ -2,6 +2,11 @@
 
 let config = {
```
`const` here as well

juanpicado commented:
@silkentrance I'm testing your branch locally, to figure out the `404`.

silkentrance commented:
@juanpicado will fix. thanks for looking into this. :+1:

juanpicado commented:
@silkentrance the `404` seems to be because in your `packages` configuration for server3 you define `pkg-nocache` while in your test you define

```javascript
const PKG_NOCACHE = 'pkg_nocache';
const PKG_NOCACHE2 = 'pkg_nocache2';
```
Thus, the uplinks are false, that drives you to 404. Then if you fix that the asserts are false, I do not know if the `PKG_NOCACHE` name is intended or not.

silkentrance commented:
@juanpicado Thanks so much for Findling

silkentrance commented:
@juanpicado Thank you so much for finding this annoying typo. I will fix that right awy incl. the issues you found in your review.

silkentrance commented:
Dunno why this was closed. Must have hit the wrong button...

silkentrance commented:
@juanpicado fixed issues and tests. rebased to master.

juanpicado commented:
Cool !! @silkentrance I'll review it asap

juanpicado commented:
Awesome work, unit testing widely complete 👍  thanks so much ! This goes to the next version.
